vendor/blob_model/source/blob_model/package.d

```diff
@@ -0,0 +1,508 @@
+     * id = ?
+     */
+    bool close(const BlobIdentifier id) @safe pure nothrow {
+        if (id.uri in blobs) {
+            blobs.remove(id.uri);
+            return true;
+        }
+        return false;
+    }
+
+    /** Get the blob matching the URI either from the VFS or the filesystem.
+     *
+     * If the blob exists in the VFS with the specific version then that is returned.
+     *
+     * Otherwise the URI is used to try and locate the blob on the filesystem.
+     *
+     * This function may throw if the URI do not exists in the internal DB and
+     * it refers to a file that do not exist on the filesystem.
+     */
+    Blob get(const Uri uri) @safe {
+        if (auto v = uri in blobs)
+            return *v;
+        return new Blob(uri, rawRead(cast(string) uri));
+    }
+
+    /// Returns: if there exists a blob with the URI.
+    bool exists(const Uri uri) @safe pure nothrow const @nogc {
+        return (uri in blobs) !is null;
+    }
+
+    /**
+     * Returns: range of the filenames in the VFS.
+     */
+    auto uris() @safe pure nothrow const @nogc {
+        return blobs.byKey;
+    }
+
+    /** Apply a stream of edits to a blob.
+     *
+     * The edits are applied starting from index zero. If there for example are
+     * two edits for the same interval the second one will be applied on top of
+     * the first one.
+     *
+     * Params:
+     * id    = blob to change
+     * edits = changes
+     */
+    bool change(const BlobIdentifier id, const(Edit)[] edits) @safe pure nothrow {
+        return change(id.uri, edits);
+    }
+
+    /// ditto
+    bool change(const BlobEdit be) @safe pure nothrow {
+        return change(be.blob.uri, be.edits);
+    }
+
+    /// ditto
+    bool change(const Uri uri, const(Edit)[] edits) @safe pure nothrow {
+        import std.algorithm : min;
+        import std.array : empty, appender;
+
+        auto blob = uri in blobs;
+        if (blob is null || edits.length == 0)
+            return false;
+
+        change(*blob, edits);
+        return true;
+    }
+}
+
+Blob change(Blob blob, const(Edit)[] edits) {
+    import std.algorithm : min;
+    import std.array : empty, appender;
+
+    foreach (const e; edits) {
+        if (e.interval.start > e.interval.end)
+            continue;
+        const start = min(e.interval.start, blob.content.length);
+        const end = min(e.interval.end, blob.content.length);
+
+        auto app = appender!(const(ubyte)[])();
+        app.put(blob.content[0 .. start]);
+        app.put(cast(const(ubyte)[]) e.content);
+        app.put(blob.content[end .. $]);
+        blob.content = app.data;
+    }
+
+    return blob;
+}
+
+/** Merge edits by concatenation when the intervals overlap.
+ *
+ * TODO: this my be a bit inefficient because it starts by clearing the content
+ * and then adding it all back. Maybe there are a more efficient way?
+ * It should at least use the allocators.
+ */
+BlobEdit merge(const Blob blob, Edit[] edits_) {
```

Comment:
@peatersson , this is what i where talking about. You can see the result in e.g. the unittests below.